contrib: add docker-machine-install-bundle.sh

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**
Added `docker-machine-install-bundle.sh`, which installs the bundle to Docker Machine instances.
This script helps developers to test the latest Docker with Swarm-mode enabled.
It is not designed to be used in production.

In future we should support [Infrakit](https://github.com/docker/infrakit) and [Testkit environment](https://github.com/docker/docker-e2e/tree/master/testkit/environment) as well.

**\- How I did it**

Copies `bundles/$VERSION/binary-{client,daemon}/{docker,dockerd,..}` to an instances's `/usr/bin` via `docker-machine ssh`.
It installs the files under `dynbinary-{client,daemon}` instead if found.

**\- How to verify it**

``` console
$ docker-machine create -d some-driver dm01
..
$ docker-machine create -d some-driver dm05
$ make binary
$ contrib/docker-machine-install-bundle.sh install dm01 dm02 dm03 dm04 dm05
Files to be installed:
bundles/1.13.0-dev/binary-daemon/dockerd
bundles/1.13.0-dev/binary-daemon/docker-proxy
bundles/1.13.0-dev/binary-daemon/docker-containerd
bundles/1.13.0-dev/binary-daemon/docker-containerd-ctr
bundles/1.13.0-dev/binary-daemon/docker-containerd-shim
bundles/1.13.0-dev/binary-daemon/docker-init
bundles/1.13.0-dev/binary-daemon/docker-runc
bundles/1.13.0-dev/binary-client/docker
dm01: stopping docker
dm02: stopping docker
dm04: stopping docker
dm03: stopping docker
dm05: stopping docker
dm05: installing docker
dm01: installing docker
dm04: installing docker
dm05: starting docker
dm01: starting docker
dm04: starting docker
dm05: done
dm01: done
dm04: done
dm03: installing docker
dm03: starting docker
dm03: done
dm02: installing docker
dm02: starting docker
dm02: done
$ eval (docker-machine env dm01)
$ docker version
```

**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

contrib: add docker-machine-install-bundle.sh

**\- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

AkihiroSuda committed:
contrib: add docker-machine-install-bundle.sh

docker-machine-install-bundle.sh installs the bundle to Docker Machine instances.

This script helps developers to test the latest Docker with Swarm-mode enabled.

Usage:

    $ docker-machine create -d some-driver dm01
    ..
    $ docker-machine create -d some-driver dm05
    $ make binary
    $ contrib/docker-machine-install-bundle.sh install dm01 dm02 dm03 dm04 dm05

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda commented:
cc @justincormack @dperny @aluzzardi

thaJeztah commented:
ping @tianon PTAL

AkihiroSuda commented:
cc @nathanleclaire

nathanleclaire commented:
Ah, very cool.  I am not maintainer on this repository, so I can't bless code for merge, but the script seems sound for the intended use case.

I have always thought a `docker-machine daemon` command to manage the daemon in a cross-init-system way would be nice, so that e.g., `docker-machine daemon restart` could replace `docker-machine ssh sudo systemctl restart docker` in this script.

nathanleclaire commented on contrib/docker-machine-install-bundle.sh:
```diff
@@ -0,0 +1,95 @@
+
+check_prereq(){
+    command -v docker-machine > /dev/null || errexit "docker-machine not installed"
+    ( tar --version | grep GNU > /dev/null ) || errexit "GNU tar not installed"
```
Good to see folks not just blindly assuming GNU tar :)

AkihiroSuda replied:
:smile:

AkihiroSuda commented:
updated PR

 * Now it can fallbacks to `service` if `systemctl` not found
 * Now it automatically detects the prefix (e.g. `/usr`)

The PR is not likely to included in 1.13 (freezed).
However, I updated PR now because I think this script is useful for developers and testers :smile:

tianon commented on contrib/docker-machine-install-bundle.sh:
```diff
@@ -0,0 +1,100 @@
+    # NOTE: `docker-machine ssh $m sh -c "foo bar"` does not work
+    #       (but `docker-machine ssh $m sh -c "foo\ bar"` works)
+    #       Anyway we avoid using `sh -c` here for avoiding confusion
+    script="if hash systemctl 2> /dev/null; then systemctl $op docker; else service docker $op; fi"
```
Can we add a fallback to `/etc/init.d/docker` so this can work on boot2docker too? :pray: :innocent:

tianon replied:
Indeed, although there is a symlink included at `/etc/init.d/docker` for better compatibility/discoverability: https://github.com/boot2docker/boot2docker/blob/bc7888a03a6cb0480bfdf31a7b615e516d0afe52/rootfs/make_iso.sh#L17 :smile: :+1:

AkihiroSuda replied:
oh, thank you for notifying that :+1: 
Updated PR

tianon reviewed (approved):
:+1:

thaJeztah reviewed (approved):
LGTM!
